metaflow/plugins/airflow/airflow_utils.py

```diff
@@ -0,0 +1,410 @@
+            "pool": str,  # the slot pool this task should run in, slot pools are a way to limit concurrency for certain tasks
+            "priority_weight": int,
+            "wait_for_downstream": bool,
+            "sla": timedelta,
+            "execution_timeout": timedelta,
+            "trigger_rule": str,
+        },
+    }
+
+    # Reference for user_defined_filters : https://stackoverflow.com/a/70175317
+    filters = dict(
+        task_id_creator=lambda v: task_id_creator(v),
+        json_dump=lambda val: json_dump(val),
+        run_id_creator=lambda val: run_id_creator(val),
+    )
+
+    def __init__(self, **kwargs):
+        self._args = kwargs
+
+    @property
+    def arguments(self):
+        return dict(**self._args, user_defined_filters=self.filters)
+
+    def serialize(self):
+        def parse_args(dd):
+            data_dict = {}
+            for k, v in dd.items():
+                # see the comment below for `from_dict`
+                if isinstance(v, dict):
+                    data_dict[k] = parse_args(v)
+                elif isinstance(v, datetime):
+                    data_dict[k] = v.isoformat()
+                elif isinstance(v, timedelta):
+                    data_dict[k] = dict(seconds=v.total_seconds())
+                else:
+                    data_dict[k] = v
+            return data_dict
+
+        return parse_args(self._args)
+
+    @classmethod
+    def deserialize(cls, data_dict):
+        def parse_args(dd, type_check_dict):
+            kwrgs = {}
+            for k, v in dd.items():
+                if k not in type_check_dict:
+                    kwrgs[k] = v
+                    continue
+                if isinstance(v, dict) and isinstance(type_check_dict[k], dict):
+                    kwrgs[k] = parse_args(v, type_check_dict[k])
+                elif type_check_dict[k] == datetime:
+                    kwrgs[k] = datetime.fromisoformat(v)
+                elif type_check_dict[k] == timedelta:
+                    kwrgs[k] = timedelta(**v)
+                else:
+                    kwrgs[k] = v
+            return kwrgs
+
+        return cls(**parse_args(data_dict, cls._arg_types))
+
+
+def _kubernetes_pod_operator_args(operator_args):
+    from kubernetes import client
+
+    from airflow.kubernetes.secret import Secret
+
+    # Set dynamic env variables like run-id, task-id etc from here.
+    secrets = [
+        Secret("env", secret, secret) for secret in operator_args.get("secrets", [])
+    ]
+    args = operator_args
+    args.update(
+        {
+            "secrets": secrets,
+            # Question for (savin):
+            # Default timeout in airflow is 120. I can remove `startup_timeout_seconds` for now. how should we expose it to the user?
+        }
+    )
+    # Below cannot be passed in dictionary form. After trying a few times it didin't work.
```

Comment:
What was the error?

Reply:
No error. It was just not recognizing the dictionary. Particularly dictionaries which have the model of `V1ObjectFieldSelector` were not being registered as environment variables.  Because of that, we were not able to pass metadata about `METAFLOW_KUBERNETES_POD_ID` etc. I also tested with a dictionary that had defaults set to the model of `V1EnvVarSource` with `V1ObjectFieldSelector` but it also didn't work. 

BTW it was easily accepting`V1EnvVar1` in JSON dictionary form but anything with `V1EnvVarSource`, didn't.